Seed package manager config

behrmann opened the pull request:
Same as #1614, which wouldn't update anymore on pushes to the branch and couldn't be reopened, because it was force pushed.

DaanDeMeyer commented on mkosi/distributions/debian.py:
```diff
@@ -159,19 +159,20 @@ def architecture(arch: Architecture) -> str:
-    state.workspace.joinpath("apt/preferences.d").mkdir(exist_ok=True)
-    state.workspace.joinpath("apt/sources.list.d").mkdir(exist_ok=True)
-    state.workspace.joinpath("apt/log").mkdir(exist_ok=True)
+    state.workspace.joinpath("pkgmngr/etc/apt").mkdir(exist_ok=True, parents=True)
```
Let's add a helper method to `MkosiState` to access this directory like we have a helper for the staging directory..

DaanDeMeyer commented on mkosi/config.py:
```diff
@@ -657,6 +658,12 @@ class MkosiConfig:
 
     preset: Optional[str]
 
+    def __post_init__(self) -> None:
+        if not self.package_manager_trees:
+            # this ugly construction is necessary to circumvent the freeze, we only care about
+            # being frozen after post init
+            object.__setattr__(self, "package_manager_trees", self.skeleton_trees)
```
Please do this using MkosiConfigSetting.default_factory

DaanDeMeyer reviewed (changes requested):
We should go further and only write files in pkgmngr if they don't already exists, even for files such as dnf.conf, so that users can override all of this if they want to.

behrmann commented:
> We should go further and only write files in pkgmngr if they don't already exists, even for files such as dnf.conf, so that users can override all of this if they want to.

Addressed the comments, but I'd like to leave this for a later follow up. This will need a bit more plumbing in the form of templating, since those configs will probably need to know the value of e.g. the workspace.

DaanDeMeyer commented on mkosi/state.py:
```diff
@@ -61,6 +62,10 @@ def workdir(self) -> Path:
     def staging(self) -> Path:
         return self.workspace / "staging"
 
+    @property
+    def pkgmngr(self) -> Path:
+        return self.workspace / "pkgmngr"
+
```
You introduced this in the wrong commit, this should be done as the first commit

DaanDeMeyer commented:
> Addressed the comments, but I'd like to leave this for a later follow up. This will need a bit more plumbing in the form of templating, since those configs will probably need to know the value of e.g. the workspace.

So at least dnf and apt support setting options via the command line. What if we move all the options that users aren't allowed to change to be set via the command line, and this would include all the options that use the workspace (and the architecture and such)? pacman also supports overriding stuff via the CLI it seems

behrmann commented on mkosi/__init__.py:
```diff
@@ -815,6 +833,7 @@ def install_unified_kernel(state: MkosiState, roothash: Optional[str]) -> None:
                 *(["--mirror", state.config.mirror] if state.config.mirror else []),
                 "--repository-key-check", yes_no(state.config.repository_key_check),
                 "--repositories", ",".join(state.config.repositories),
+                *flatten([("--package-manager-tree", format_source_target(s, t)) for s, t in state.config.package_manager_trees]),
```
This would be more similar to the rest, but I think
```python
*(["--package-manager-tree", ",".join(format_source_target(s, t) for s, t in state.config.package_manager_trees] if state.config.package_manager_trees else [])
```
is less readable than the `flatten`.

DaanDeMeyer replied:
You should be able to get away with 

```python
"--package-manager-tree", ",".join(...)
```

behrmann replied:
Exchanged it for that.

DaanDeMeyer commented on mkosi/distributions/arch.py:
```diff
@@ -82,8 +86,16 @@ def install_packages(cls, state: MkosiState, packages: Sequence[str], apivfs: bo
-            for d in state.config.repo_dirs:
-                f.write(f"Include = {d}/*\n")
+            confsnippetsdir = state.pkgmngr / "etc/pacman.d/conf/"
+            if any(confsnippetsdir.glob("*")):
```
iterdir() instead of glob()?

behrmann replied:
Sure, made it a bit more robust, too.

DaanDeMeyer commented on mkosi/distributions/debian.py:
```diff
@@ -245,10 +246,3 @@ def install_apt_sources(state: MkosiState, repos: Sequence[str]) -> None:
             for repo in repos:
                 f.write(f"{repo}\n")
-
-    # Already contains a merged tree of repo_dirs after setup_apt
-    for src in state.workspace.joinpath("apt/sources.list.d").iterdir():
-        dst = state.root.joinpath("etc/apt/sources.list.d", src.name)
-        if dst.exists():
-            continue
-        shutil.copyfile(src, dst)
```
Can we just drop this? This is writing to the root, which we don't do anywhere else?

behrmann replied:
Thought about this already and wasn't entirely sure, but it shouldn't be a problem and the file can be easily passed in via some tree. Added it as an extra commit in case it needs reverting.

DaanDeMeyer reviewed (changes requested):
Almost there. We can either do the CLI stuff in this PR or another one

DaanDeMeyer reviewed (changes requested):
Almost there. We can either do the CLI stuff in this PR or another one

DaanDeMeyer commented on mkosi/distributions/debian.py:
```diff
@@ -124,7 +124,6 @@ def install_packages(cls, state: MkosiState, packages: Sequence[str], apivfs: bo
         setup_apt(state, cls.repositories(state))
         invoke_apt(state, "update", apivfs=False)
         invoke_apt(state, "install", packages, apivfs=apivfs)
-        install_apt_sources(state, cls.repositories(state, local=False))
```
The thing is, if we drop this, the user has to write the file himself, which is kind of a bad user experience unlike the other distros where you can just install a package, so I'd keep this intact for now

behrmann replied:
Dropped the commit again.

DaanDeMeyer commented on mkosi/distributions/debian.py:
```diff
@@ -216,7 +217,7 @@ def setup_apt(state: MkosiState, repos: Sequence[str]) -> None:
                 continue
             if src.suffix in (".list", ".sources"):
-                shutil.copyfile(src, state.workspace.joinpath("apt/sources.list.d", src.name))
+                shutil.copyfile(src, state.workspace.joinpath("pkgmngr/etc/apt/sources.list.d", src.name))
```
`state.pkgmngr`

behrmann replied:
same

DaanDeMeyer commented on mkosi/distributions/fedora.py:
```diff
@@ -172,10 +174,9 @@ def invoke_dnf(
         "--setopt=keepcache=1",
         "--setopt=install_weak_deps=0",
         f"--setopt=cachedir={state.cache_dir}",
-        f"--setopt=reposdir={' '.join(str(p) for p in state.config.repo_dirs)}",
```
We drop this but never set it again to the path in the pkgmngr tree?

behrmann replied:
That must have gotten lost during a rebase.

DaanDeMeyer reviewed (changes requested):
Found a few more things

behrmann commented on mkosi/distributions/arch.py:
```diff
@@ -84,6 +84,18 @@ def install_packages(cls, state: MkosiState, packages: Sequence[str], apivfs: bo
                     )
                 )
 
+            confsnippetsdir = state.pkgmngr / "etc/pacman.d/conf/"
+            # glob is used instead of iterdir here, because it doesn't throw an error if
+            # confsnippetsdir does not exist.
+            if any(f.is_file() for f in confsnippetsdir.glob("*")):
+                f.write(
+                    dedent(
+                        f"""\
+
+                        Include = {state.pkgmngr}/etc/pacman.d/conf/*
+                        """
+                    )
+                )
```
That doesn't work, pacman complains about no such file existing.

behrmann commented on mkosi/distributions/debian.py:
```diff
@@ -247,7 +248,7 @@ def install_apt_sources(state: MkosiState, repos: Sequence[str]) -> None:
 
     # Already contains a merged tree of repo_dirs after setup_apt
-    for src in state.workspace.joinpath("apt/sources.list.d").iterdir():
+    for src in state.workspace.joinpath("pkgmngr/etc/apt/sources.list.d").iterdir():
```
I don't see it and GitHub showed the hunk as outdated from the start.

behrmann commented on mkosi/distributions/debian.py:
```diff
@@ -207,7 +208,7 @@ def setup_apt(state: MkosiState, repos: Sequence[str]) -> None:
     )
 
-    with state.workspace.joinpath("apt/sources.list").open("w") as f:
+    with state.workspace.joinpath("pkgmngr/etc/apt/sources.list").open("w") as f:
```
same
